### ISSUE TITLE
Searching on child properties

### ISSUE THREAD

**Issue by Hyra** [markhuot/craftql#187]
Heya,

I'm trying to figure out how you can search on child properties. I've seen the 'search' param, which works fine when the subfield is just text. But if it's a nested object like a list or entries I can't seem to figure out how to limit the results.

My current query:

<img width="723" alt="query" src="https://user-images.githubusercontent.com/380836/48247121-70459800-e3f2-11e8-810b-e161a6d45337.png">

So essentially I want to only query the entries that match a certain category id by searching in the categoryField

Any suggestions?

**Comment by markhuot** [comment]
You should be able to use `search:` the same way you would in a `.twig` template. So the following _should_ work,

```graphql
{
  entries(section:ledenVoordelen, search:”categoryField:21”) {
    # ...
  }
}
```

**Comment by Hyra** [comment]
Hmm, I did try that, but when I do it gives me zero results.

Without search parameter:

<img width="1350" alt="screenshot 2018-11-09 13 49 33" src="https://user-images.githubusercontent.com/380836/48263417-5625ad00-e426-11e8-9c32-fb582f75af04.png">

With search:

<img width="1355" alt="screenshot 2018-11-09 13 50 15" src="https://user-images.githubusercontent.com/380836/48263442-689fe680-e426-11e8-9179-b3a473c58524.png">

So close ..

**Comment by Hyra** [comment]
Could it be because the data is inside a nested structure? searching on top level properties works fine, but the id i'm searching for is inside the 'voordeelCategorie'  structure.

**Comment by narrishinsd** [comment]
@Hyra , it seems likely, since I think your query only identifies the top level of voordeelCatagorie.

To query deeper, you need recursion. In GraphQL that’s done manually – there’s  no recursive ability.

So here’s copy-paste of an example of something similar for entries with Structure children. Fragments allow constructing the depth. It's very late, so apologies for the over-complication.

The 'manual recursion' occurs in the second fragment, `recursive2Nav`. You can see it's total three deep here, and how to extend if you need more.

You can adapt and see if search is smart enough to follow down into your query.  

We may hope….
Clive

```
fragment navFields on PagesPage {
  title
  level
  uri
  id
  children (idNot:$idNot) {
    id
  }
}

fragment recursive2Nav on PagesPage {
  ...navFields
  children (idNot:$idNot) {
    ...navFields
    children (idNot:$idNot) {
      ...navFields
    }
  }
}

query page($id: [Int], $site: String, $idNot: [Int]) {
  nav: entries(section: pages, level: 1, idNot: $idNot) {
    ...recursive2Nav
  }
  pageContent: entries(id: $id, site: $site, section: pages) {
    ... on PagesPage {
      title
      uri
      id
      postDate @date(as: "j F Y")
      mainContent: pageMContent {
        type: __typename
        ... on PageMContentTextArea {
          textArea {
            totalPages
            content
          }
        }
        ... on PageMContentAccordion {
          title: accordionTitle
          content: accordionContent
        }
      }
    }
  }
}
```

**Comment by narration-sd** [comment]
@Hyra , it seems likely, since I think your query only identifies the top level of voordeelCatagorie.

To query deeper, you need recursion. In GraphQL that’s done manually – there’s  no recursive ability, by intention.

So here’s copy-paste of an example of something similar for entries with Structure children. Fragments allow constructing the depth. It's very late, so apologies for the over-complication.

The 'manual recursion' occurs in the second fragment, `recursive2Nav`. You can see it's total three deep here, and how to extend if you need more.

You can adapt and see if search is smart enough to follow down into your query. 

We may hope….
Clive

```
fragment navFields on PagesPage {
  title
  level
  uri
  id
  children (idNot:$idNot) {
    id
  }
}

fragment recursive2Nav on PagesPage {
  ...navFields
  children (idNot:$idNot) {
    ...navFields
    children (idNot:$idNot) {
      ...navFields
    }
  }
}

query page($id: [Int], $site: String, $idNot: [Int]) {
  nav: entries(section: pages, level: 1, idNot: $idNot) {
    ...recursive2Nav
  }
  pageContent: entries(id: $id, site: $site, section: pages) {
    ... on PagesPage {
      title
      uri
      id
      postDate @date(as: "j F Y")
      mainContent: pageMContent {
        type: __typename
        ... on PageMContentTextArea {
          textArea {
            totalPages
            content
          }
        }
        ... on PageMContentAccordion {
          title: accordionTitle
          content: accordionContent
        }
      }
    }
  }
}
```

**Comment by narration-sd** [comment]
Hmm. i may have given you a wrong steer...and in any case, this is going to take @markhuot having a further look, I think.

I hadn't used `search:` before, so constructed a model and tried things -- see the image.

But:
- I couldn't get `search:` to find on any field, not even titles; just got nothing selected whatever I tried, including quotings, using a query variable rather than inline, etc.. 
- so unless I'm making an obvious mistake, `search;` may not be working at present on dev-master.
- but also, I don't think you actually need the recursion I put forward, unless you want to actually present it. That's because Categories are returned in a simple array on the entry, which  the image shows. Only their level (and parent/child) indicates hierarchy, so it's not involved in search:

There's also a question about `search:` syntax or positioning -- in this case, say you wanted to find among category titles; how then is that differentiated from entry titles, esp. since the handle is immutable in Craft?

Will be interested what the answer/s turn out here. 

![image](https://user-images.githubusercontent.com/247945/48305859-25597c80-e4e7-11e8-84dd-408d1f139324.png)

**Comment by narration-sd** [comment]
sorry for bad habit  of editing  after posting -- the result should be a bit clearer when you read it on Github...

**Comment by Hyra** [comment]
Not a problem. Really "happy" I'm not the only one, as I ran out of clues.

Following up on your last example, you probably get autocompletion on "dealCategory" inside the entries parameters, which would indicate one can query against it. Just not sure what the value would be. "14" in your case would make sense, but it will probably not return any results.

**Comment by Hyra** [comment]
My problem with this accidentally got solved by issue #188 :)

I can now query for entries which have a specific category:

```
query GetEntries {
  categoriesConnection (slug: "some-slug") {
    edges {
      relatedEntries (section: ledenVoordelen) {
        entries {
          id
          title
          ... on LedenVoordelen {
            voordeelcategorie {
              title
            }
          }
        }
      }
    }
  }
}
```


Question remains how one would do this when you want to query on a value inside a Matrix field for instance, but for now I got this one solved!

**Comment by narration-sd** [comment]
@Hyra  Stef, had no time until a few moments today, but interested. I see the route you took, and I guess it's a thing to learn about how to think about GraphQl.

I went back and looked at how the search should work  -- it does. So this is another way of specifying which category you are interested in. No doubt the slug query is more efficient, vs. convenience.

Then an example is attached, also with hierarchy. It isn't completely correct because in this edges formulation, I couldn't find a place acceptable to specify the level, which should be : 1 before the recursion is called, so that you wouldn't get the extra output  lines double listing the cheeses.

Maybe @markhuot has an idea for this, when he has time.

Anyway, glad you got  your  project moving. And the Matrix field query...have to go back to tasks here, but would think a MatrixConnection would show up when a Matrix field was present, and then should be (kind of) similar approach, as a guess.

![kaas-platten](https://user-images.githubusercontent.com/247945/48645862-ce96eb80-e99b-11e8-8bab-9d57d5d51e2a.png)
